src/core/aio.c

```diff
@@ -523,16 +527,61 @@ nni_aio_expire_add(nni_aio *aio)
 		// And, as we are the latest, kick the thing.
 		nni_cv_wake(&aio->a_expire_q->eq_cv);
 	}
+	*/
+
+	nni_aio_expire_q * eq = aio->a_expire_q;
+	if (eq->eq_len > eq->eq_cap * 2 / 3) {
+		nni_aio ** new_list =
+		    nni_zalloc(eq->eq_cap * 2 * sizeof(nni_aio *));
+		for (uint32_t i = 0; i < eq->eq_len; i++) {
+			new_list[i] = eq->eq_list[i];
+		}
+		nni_free(eq->eq_list, eq->eq_cap * sizeof(nni_aio *));
+		eq->eq_list = new_list;
+		eq->eq_cap *= 2;
+	}
+	eq->eq_list[eq->eq_len++] = aio;
+	if (eq->eq_len == 1) {
+		nni_cv_wake(&eq->eq_cv);
+	}
+}
+
+static void
+nni_aio_expire_rm(nni_aio *aio)
+{
+	uint32_t          aio_idx = 0;
+	nni_aio_expire_q *eq      = aio->a_expire_q;
+	for (aio_idx = 0; aio_idx < eq->eq_len; aio_idx++) {
+		if (aio == eq->eq_list[aio_idx]) {
+			for (uint32_t i = aio_idx + 1; i < eq->eq_len; i++) {
+				eq->eq_list[i - 1] = eq->eq_list[i];
+			}
+			eq->eq_len --;
+			break;
+		}
+	}
+
```

Comment:
i'd be reasonably inclined *not* to automatically shrink the list... or to have the size at which we shrink be a tunable.  I would not ever expect the size of this array to be a serious problem -- for example if you have 1M items outstanding for expiration, you're only going to need 8MB.  That's an extreme number of aios (outside the real world for certain!) while still only needing a relatively modest queue size.

Avoiding the churn on the allocator is a very worthwhile thing to do.

Reply:
Hi Garrett, Totally agree with you. Hence, we never shrink the list or leave a tunable CMAKE parameter like NNG_MAX_AIO_LIST to users so that they can have control over the exact memory is consumed?